### PR TITLE
Support for ListObjectParts in azure-gateway

### LINKED ISSUE #5169: [feature] Azure Gateway: Implement ListObjectParts
## Expected Behavior

`ListObjectParts` returns a list of parts which can be used to generate a subsequent `CompleteMultipartUpload` request

## Current Behavior

`ListObjectParts` always returns an empty list, as per Known Limitations list in Azure Gateway docs.

## Possible Solution

The `ListObjectParts` stub method in Azure Gateway has the comment `ListObjectParts - Use Azure equivalent GetBlockList`, suggesting that it may be possible to implement `ListObjectParts` using existing Azure API methods. 

## Context

While attempting to set up a `tusd` server which stores uploaded files on Azure, I came across Minio Azure Gateway and thought it seemed like a neat way to achieve this as `tusd` supports S3 (but doesn't currently support Azure).  However, `tusd` currently fails to complete the multipart upload as it relies on S3 to provide the list of parts.  I [raised an issue](https://github.com/tus/tusd/issues/156) with `tusd` to see if it was possible to track the parts there, but its architecture makes it infeasible to do so - especially if `tusd` is scaled horizontally (i.e. there are multiple instances of `tusd` potentially handling different chunks of the same upload).

## Your Environment

* Version used (`minio version`): RELEASE.2017-10-27T18-59-02Z
* Environment name and version (e.g. nginx 1.9.1): `minio/minio` Docker image

### PR DESCRIPTION
**Pull request by krishnasrinivas**
fixes #5169

## How Has This Been Tested?

manual. mint functional test should be added: https://github.com/minio/mint/issues/223

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [ ] All new and existing tests passed.

### CODE REVIEW

File: cmd/gateway-azure.go [hunk 1]
```diff
@@ -738,7 +739,76 @@ func (a *azureObjects) ListObjectParts(bucket, object, uploadID string, partNumb
...
+	// Build a sorted list of parts and return the requested entries.
+	partsMap := make(map[string]PartInfo)
+	for _, block := range resp.UncommittedBlocks {
+		var partNumber int
+		var parsedUploadID string
+		var md5Hex string
+		if partNumber, _, parsedUploadID, md5Hex, err = azureParseBlockID(block.Name); err != nil {
+			return result, azureToObjectError(traceError(errUnexpected), bucket, object)
+		}
+		partNumberStr := strconv.Itoa(partNumber)
```
Comment (krisis) [inline cmd/gateway-azure.go]: can't we use a `map[int]PartInfo` and avoid converting `partNumber` to `string`?

File: cmd/gateway-azure.go [hunk 2]
```diff
@@ -738,7 +739,76 @@ func (a *azureObjects) ListObjectParts(bucket, object, uploadID string, partNumb
...
+	}
+	for partsCount < maxParts && i < len(parts) {
+		result.Parts = append(result.Parts, parts[i])
+		i++
+		partsCount++
+	}
+
+	if i < len(parts) {
+		result.IsTruncated = true
+		if partsCount != 0 {
```
Comment (krisis) [inline cmd/gateway-azure.go:807]: will `partsCount` ever be zero here?

**Review by krisis — CHANGES_REQUESTED** [review]

File: cmd/gateway-azure.go (hunk 1 above)
Comment (krishnasrinivas) [reply cmd/gateway-azure.go]: fixed it

File: cmd/gateway-azure.go (hunk 2 above)
Comment (krishnasrinivas) [reply cmd/gateway-azure.go:807]: Yes, if a misbehaving client does:
```
GET /test/testobject?max-parts=0&part-number-marker=0&uploadId=0239618213468449 HTTP/1.1
Host: localhost:9000

```

**Review by balamurugana — APPROVED** [review]
LGTM

**Review by krisis — APPROVED** [review]

**Pull request merged**